File: Tree-shaped Deep Forest/Binary Tree/Q3. Longest Univalue Path/Q3_LongestUnivaluePath.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def longestUnivaluePath(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        self.max = 0  # stores maximum nodes in same-value path

        def dfs(node):
            if not node:
                return 0  # base case

            left = dfs(node.left)    # longest path from left subtree
            right = dfs(node.right)  # longest path from right subtree

            left_arrow = 0
            right_arrow = 0

            # extend path to left if value matches
            if node.left and node.left.val == node.val:
                left_arrow = left

            # extend path to right if value matches
            if node.right and node.right.val == node.val:
                right_arrow = right

            # update global max path passing through current node
            self.max = max(self.max, left_arrow + right_arrow + 1)

            # return longest single direction path upward
            return max(left_arrow, right_arrow) + 1

        dfs(root)  # start DFS traversal

        # convert node count to edge count
        return self.max - 1 if self.max > 0 else 0

File: Tree-shaped Deep Forest/Binary Tree/Q3. Longest Univalue Path/Q3_LongestUnivaluePath.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int max = 0; // stores maximum nodes in same-value path

    public int longestUnivaluePath(TreeNode root) {
        maxPath(root); // start DFS traversal
        return max > 0 ? max - 1 : 0; // convert nodes count to edges count
    }

    private int maxPath(TreeNode node) {
        if (node == null) return 0; // base case

        int left = maxPath(node.left);   // longest path from left subtree
        int right = maxPath(node.right); // longest path from right subtree

        int leftArrow = 0, rightArrow = 0; // valid paths with same value

        // extend path from left if value matches
        if (node.left != null && node.left.val == node.val) {
            leftArrow = left;
        }

        // extend path from right if value matches
        if (node.right != null && node.right.val == node.val) {
            rightArrow = right;
        }

        // update global maximum path passing through this node
        max = Math.max(max, leftArrow + rightArrow + 1);

        // return longest single direction path upward
        return Math.max(leftArrow, rightArrow) + 1;
    }
}